src/modules/silentpayments/main_impl.h

```diff
@@ -0,0 +1,674 @@
+    const secp256k1_silentpayments_label_lookup label_lookup,
+    const void *label_context
+) {
+    secp256k1_scalar t_k_scalar, rsk_scalar;
+    secp256k1_ge label_ge, recipient_spend_pubkey_ge, A_sum_ge;
+    secp256k1_pubkey A_sum;
+    secp256k1_xonly_pubkey P_output_xonly;
+    unsigned char shared_secret[33];
+    const unsigned char *label_tweak = NULL;
+    size_t i, k, n_found, found_idx;
+    int found, combined;
+    int ret = 1;
+
+    /* Sanity check inputs */
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(found_outputs != NULL);
+    ARG_CHECK(n_found_outputs != NULL);
+    ARG_CHECK(tx_outputs != NULL);
+    ARG_CHECK(n_tx_outputs > 0);
+    ARG_CHECK(recipient_scan_key != NULL);
+    ARG_CHECK(public_data != NULL);
+    ARG_CHECK(recipient_spend_pubkey != NULL);
+    /* Passing a context without a lookup function is non-sensical */
+    if (label_context != NULL) {
+        ARG_CHECK(label_lookup != NULL);
+    }
+    /* TODO: do we need a _cmov call here to avoid leaking information about the scan key?
+     * Recall: a scan key isnt really "secret" data in that leaking the scan key will only leak privacy
+     * In this respect, a scan key is functionally equivalent to an xpub
+     */
+    /* If there is something wrong with the recipient scan key, recipient spend pubkey, or the public data, then return */
+    ret = secp256k1_scalar_set_b32_seckey(&rsk_scalar, recipient_scan_key);
+    /* TODO: only fails in case of invalid key */
+    secp256k1_declassify(ctx, &ret, sizeof(ret));
+    if (!ret) {
+        /* consider clearing */
+        return 0;
+    }
+    if (!secp256k1_silentpayments_recipient_public_data_load_pubkey(ctx, &A_sum, public_data)) {
+        return 0;
+    }
+    if (!secp256k1_pubkey_load(ctx, &A_sum_ge, &A_sum)) {
+        return 0;
+    }
+    if (!secp256k1_pubkey_load(ctx, &recipient_spend_pubkey_ge, recipient_spend_pubkey)) {
+        return 0;
+    }
+    combined = (int)public_data->data[0];
+    if (!combined) {
+        unsigned char input_hash[32];
+        secp256k1_scalar input_hash_scalar;
+        int overflow = 0;
+
+        secp256k1_silentpayments_recipient_public_data_load_input_hash(input_hash, public_data);
+        secp256k1_scalar_set_b32(&input_hash_scalar, input_hash, &overflow);
+        if (overflow) {
+            return 0;
+        }
+        secp256k1_scalar_mul(&rsk_scalar, &rsk_scalar, &input_hash_scalar);
+    }
+    secp256k1_silentpayments_create_shared_secret(ctx, shared_secret, &rsk_scalar, &A_sum_ge);
+
+    found_idx = 0;
+    n_found = 0;
+    k = 0;
+    while (1) {
+        secp256k1_ge P_output_ge = recipient_spend_pubkey_ge;
+        /* Calculate t_k = hash(shared_secret || ser_32(k)) */
+        secp256k1_silentpayments_create_t_k(&t_k_scalar, shared_secret, k);
+
+        /* Calculate P_output = B_spend + t_k * G
+         * This can fail if t_k overflows the curve order, but this is statistically improbable
+         */
```

Comment:
This comment doesn't seem to be true. This only fails if `t_k = -b_spend` where `b_spend*G = B_spend`.
```suggestion
        /* Calculate P_output = B_spend + t_k * G
         * This can fail if t_k equals the negation of the DLog of P_output, but t_k
         * is the output of a hash function.
         */
        ret = secp256k1_eckey_pubkey_tweak_add(&P_output_ge, &t_k_scalar);
```
Also, aren't we calculating `B_spend` and not `P_output`?

Reply:
The comment is indeed incorrect, since we aren't ever checking if the output of the hash function (`t_k`) overflows the curve order.

I can have `_create_t_k` fail if `t_k_scalar` overflows, although this introduces a branch that we are unable to test (and is statistically improbable). This would, however, make the function more strictly adhere to the [specification](https://github.com/bitcoin/bips/blob/master/bip-0352.mediawiki#creating-outputs)